refactor(core): Add four tests and fix code to make tests pass.

iteriani opened the pull request:
The first test asserts that bubbling does not work right now.

The second asserts that stopPropagation works, which should pass when test #1 passes too.

The third test asserts properties about the events passed to the event handler.

THe fourth test asserts that mouse events do not translate to jsaction nor help emit the jsaction binary. This required a change in code to make this pass.

The change looks like a lot because the indentation was fixed, but look at "bubbling behavior" and "should not replay mouse events"

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

iteriani committed:
refactor(core): Add four tests and fix code to make tests pass.

The first test asserts that bubbling does not work right now.

The second asserts that stopPropagation works, which should pass when test #1 passes too.

The third test asserts properties about the events passed to the event handler.

THe fourth test asserts that mouse events do not translate to jsaction nor help emit the jsaction binary. This required a change in code to make this pass.

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -37,88 +37,82 @@ function hasJSActionAttrs(content: string) {
+  const originalWindow = globalThis.window;
+
+  beforeAll(async () => {
+    globalThis.window = globalThis as unknown as Window & typeof globalThis;
```
This seems a bit odd? What's going on here?

iteriani replied:
It is odd but basically this is an oddity in that the window object at test start time is not the same window object during the test.

thePunderWoman replied:
This could be due to dom emulation. The window object at the start of the test would be the one created by Domino on the server side, and then later once the client is replaying, you get the native window object.

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -37,88 +37,82 @@ function hasJSActionAttrs(content: string) {
+    const {
+      enableEventReplay = true,
+      hydrationDisabled,
+      doc = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`,
```
move to constant?

iteriani replied:
done

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -37,88 +37,82 @@ function hasJSActionAttrs(content: string) {
+  function render(doc: Document, html: string) {
+    renderHtml(doc, html);
+    globalThis.document = doc;
+    const scripts = doc.getElementsByTagName('script');
```
Can you add an id to the script so you don't have to search through all scripts on the page?

iteriani replied:
I'm being a bit more forward looking in this case, since we may expect more than one script that does this.

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
-        expect(clickSpy).toHaveBeenCalled();
-        expect(focusSpy).toHaveBeenCalled();
-      });
+    const docContents = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
This is the default isn't it?

iteriani replied:
yeah

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+    const docContents = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
+    const html = await ssr(SimpleComponent, {doc: docContents});
+    const ssrContents = getAppContents(html);
+    expect(ssrContents).toContain(
```
Not sure we need to assert the literal text content of the script. Isn't testing behavior enough?

iteriani replied:
sure

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+    render(doc, ssrContents);
+    const el = doc.getElementById('click-element')!;
+    const button = doc.getElementById('focus-target-element')!;
+    const clickEvent = new CustomEvent('click', {bubbles: true});
```
Leave a comment that this test runs in node and cannot use normal browser methods.

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+    const button = doc.getElementById('focus-target-element')!;
+    const clickEvent = new CustomEvent('click', {bubbles: true});
+    el.dispatchEvent(clickEvent);
+    const focusEvent = new CustomEvent('focus');
```
Leave a comment about node as well

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
-        expect(el.hasAttribute('jsaction')).toBeFalse();
-        expect((el.firstChild as Element).hasAttribute('jsaction')).toBeFalse();
-      });
+    const docContents = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
This is the default I believe

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+    render(doc, ssrContents);
+    const el = doc.getElementById('1')!;
+    expect(el.hasAttribute('jsaction')).toBeTrue();
+    expect((el.firstChild as Element).hasAttribute('jsaction')).toBeTrue();
```
getElementById('2') since you have the id anyways

tbondwilkinson replied:
And assign it so that you can get it below without having to do firstChild again

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+      `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}` +
+      `<app ngCspNonce="{{nonce}}"></app></body></html>`;
+    const html = await ssr(SimpleComponent, {doc});
+    expect(getAppContents(html)).toContain('<script nonce="{{nonce}}">window.__jsaction_bootstrap');
```
Can you just check that the script tag has a nonce attribute that matches? I don't necessarily think we need the whole text check.

Also if you give script an id you can use that instead of doing the text search

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+      class SimpleComponent {
+        onClick = onClickSpy;
+      }
+      const docContents = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
This is the default

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+      render(doc, ssrContents);
+      resetTViewsFor(SimpleComponent);
+      const bottomEl = doc.getElementById('bottom')!;
+      bottomEl.click();
```
Wait I thought click didn't work? You used `CustomEvent` for it earlier in this test

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+        hydrationFeatures: [withEventReplay()],
       });
+      appRef.tick();
+      // This is a bug
```
Say what the bug specifically is.

The expected behavior is to have two calls, one for bottom, one for top.

This would be more self-evident if you used two separate spys.

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+      bottomEl.click();
+      expect(onClickSpy).toHaveBeenCalledTimes(2);
+    });
+    it('should not propagate events if stopPropagation is called', async () => {
```
add newlines between test cases

thePunderWoman replied:
➕ 1

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+        }
+      }
+      const onClickSpy = spyOn(SimpleComponent.prototype, 'onClick').and.callThrough();
+      const docContents = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
this is the default

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+          currentEvent = event;
+        }
+      }
+      const docContents = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
this is the default

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+        onClick() {}
+      }
+
+      const doc = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
still the default I believe

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+      })
+      class SimpleComponent {}
+
+      const doc = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
same

iteriani replied:
done

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+        doThing() {}
+      }
+
+      const doc = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
same

iteriani replied:
done

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+        onClick() {}
+      }
+
+      const doc = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
same

iteriani replied:
done

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+        onClick() {}
+      }
+
+      const doc = `<html><head></head><body>${EVENT_DISPATCH_SCRIPT}<app></app></body></html>`;
```
same

iteriani replied:
done

tbondwilkinson commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +121,289 @@ describe('event replay', () => {
+      // event contract setup goes second (since it uses some code from
+      // the inlined script).
+      expect(ssrContents).toContain(
+        `<script type="text/javascript" id="ng-event-dispatch-contract"></script>` +
```
text contains is brittle, there's probably a better way here as well

iteriani replied:
maybe but not the point of this PR

iteriani committed:
refactor(core): Simplify event handler extraction logic.

This reuses information already recorded during hydration to
remove jsaction attributes to also stash event handlers. This avoids
a tree walk and looku.

iteriani committed:
fixup! refactor(core): Add four tests and fix code to make tests pass.

thePunderWoman commented on packages/platform-server/test/event_replay_spec.ts:
```diff
@@ -127,172 +118,280 @@ describe('event replay', () => {
+      bottomEl.click();
+      expect(onClickSpy).toHaveBeenCalledTimes(1);
+    });
+    it('should not have differences in event fields', async () => {
```
Same here with the line break, please.

iteriani replied:
done

thePunderWoman reviewed (approved):
reviewed-for: fw-core, fw-platform-server

iteriani committed:
fixup! refactor(core): Add four tests and fix code to make tests pass.

atscott commented:
This PR was merged into the repository by commit 3ad120dfe84eb80e4c4d7fcd912fbe26058b4233.
